### ISSUE TITLE
Inconsistency in MirroredStrategy evaluation results for batch dependent metrics

### ISSUE THREAD

**Issue by zaccharieramzi** [tensorflow/tensorflow#39268]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux Ubuntu 16.04
- TensorFlow installed from (source or binary): colab (and from pip locally)
- TensorFlow version (use command below): 2.2.0-rc4 (but also locally on 2.1)
- Python version: 3.6.9 (and 3.6.8 locally) 
- CUDA/cuDNN version: 10.1
- GPU model and memory: Quadro P50000 locally, colab otherwise

**Describe the current behavior**

When I use `MirroredStrategy` for model evaluation with whole-batch dependent metrics, there is some inconsistency in the metrics returned.
This has to do with the fact that the batches are separated before being sent to the different devices and the metrics are computed on each device before being averaged on the master device.

**Describe the expected behavior**

I would like the metrics evaluation to be independent of whether I use `MirroredStrategy` or not.

**Standalone code to reproduce the issue**
https://colab.research.google.com/drive/1_Lohjz7qSjF7cGKCMO7F7-lIBDMTYJw5?usp=sharing

You can test that the metric computation is otherwise consistent by changing the flag to `False`.

```python
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense


# logical device separation
physical_devices = tf.config.list_physical_devices('GPU')

if True:
  tf.config.set_logical_device_configuration( 
      physical_devices[0], 
      [tf.config.LogicalDeviceConfiguration(memory_limit=8000), 
       tf.config.LogicalDeviceConfiguration(memory_limit=8000)])


# my full batch dependent loss
def my_loss(y_true, y_pred):
    return tf.reduce_max(tf.abs(y_true - y_pred))

# my toy model
mirrored_strategy = tf.distribute.MirroredStrategy(cross_device_ops=tf.distribute.ReductionToOneDevice())
with mirrored_strategy.scope():
    model_distributed = Sequential(Dense(10))
    model_distributed.compile(loss=my_loss)

# my toy data
x = tf.random.normal([32, 10])
y = tf.random.normal([32, 10])

# my experiments
metrics = model_distributed.evaluate(x, y)
print(metrics)

y_pred = model_distributed.predict(x)
print(my_loss(y, y_pred))
```

**Other info / logs**

A sample output from the previous code would for example be:
```
1/1 [==============================] - 0s 1ms/step - loss: 4.1214
4.121423721313477
tf.Tensor(4.1544814, shape=(), dtype=float32)
```

An obvious solution is to not use `evaluate` but `predict` and iterate myself over the dataset (in my real use case I use a dataset and NCCL) computing the metrics myself. But I am then losing some nice properties of `evaluate` like the callbacks and I have to compute manually potentially a range of metrics.

Maybe this isn't a bug but in which case it would be nice to have a warning in the docs. I also would like to know if there is a way to still use `evaluate` maybe with a custom `cross_device_ops`.

**Comment by ravikyram** [comment]
I have tried in colab with TF 2.2-rc3 and was able to reproduce the issue.Please, find the gist [here](https://colab.sandbox.google.com/gist/ravikyram/dfeba6541562b066b70473c0e56a0e3d/untitled866.ipynb).Thanks!

**Comment by guptapriya** [comment]
Whole batch dependent metrics are possible to implement if you create a custom metric: https://www.tensorflow.org/guide/keras/train_and_evaluate#custom_metrics. 

This is how the standard metrics are implemented as well. In this you can define how each replica accumulates the results independently on their local batch, and how those values should be combined to give the final result. However, I think the only reduction types available currently are SUM and MEAN. We don't support MAX/MIN yet which is what you probably need for this case? 

You can check our the inbuilt metrics implementations to see how they're made to work when used w distribution strategy: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/metrics.py

**Comment by zaccharieramzi** [comment]
Thanks for this info ! Can you point me to a specific metric which handles the distribution strategy, I didn't find one right away?

Regarding the custom metric, I am not sure how the whole state thing works in the evaluation mode. I guess when fitting it just resets at every epoch, but what about when evaluating? Does it reset at every batch?

Indeed I need something which can handle the max and the mean. Basically what I am trying to do is compute the PSNR of a whole volume given as a batch. The [formula](https://en.wikipedia.org/wiki/Peak_signal-to-noise_ratio) for the volume PSNR is `10 * log_10 (max(volume)**2 / mse(volume))`. Basically the `mse` over the volume could just be a mean reduction but then I need to compute the max over the volume and compute this `10 * log_10` operation.

**Comment by guptapriya** [comment]
All the inbuilt metrics handle distribution strategy be defining update_state and result methods. There are wrappers for those in [metric_utils](https://github.com/tensorflow/tensorflow/blob/5d49dc5526324443931a33cc84d66c8bcae9cea2/tensorflow/python/keras/utils/metrics_utils.py#L63) which get used and handle the logic to make those methods work in distribution strategy context. 

But all of them rely on averaging the underlying metric variables across replicas.. we don't support reduce_max yet across replicas. 

Re: state, metrics reset every epoch so I think in evaluation, they will not reset at all? @pavithrasv can confirm.

**Comment by pavithrasv** [comment]
That's correct, that will not reset in evaluation.

**Comment by zaccharieramzi** [comment]
Ok too bad for the reduce_max. Maybe we should close this for the time being.

**Comment by sushreebarsa** [comment]
Was able to replicate the issue with TF 2.5,please find the gist[ here](https://colab.research.google.com/gist/sushreebarsa/0e0f31dcdca22403f37732eee3ec276e/untitled866.ipynb#scrollTo=wjlNSm5QNx7d) ..Thanks!

**Comment by mohantym** [comment]
Was able to replicate this issue in [TF 2.7](https://colab.research.google.com/gist/mohantym/21cccf6c287482f3771d35e84271694a/untitled866.ipynb#scrollTo=wjlNSm5QNx7d) too .Thanks!
